File: hbTapi Components/BCB Samples/Enterprise Edition/CallData/uMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "uMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "hbTapi"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TForm1::ComboBoxDeviceChange(TObject *Sender)
{
  hbTapiLine1->Active = false;
  hbTapiLine1->DeviceID = ComboBoxDevice->ItemIndex-1;
  ComboBoxAddress->Items->Assign(hbTapiLine1->AddressList);
  ComboBoxAddress->ItemIndex = 0;
  if (hbTapiLine1->Available)
  {
    if (ComboBox_Privileges->Text.Pos("AutoSelect") > 0)
      hbTapiLine1->Privileges->AutoSelect = true;
    else
    {
      hbTapiLine1->Privileges->None = true;
      hbTapiLine1->Privileges->AutoSelect = false;
      if (ComboBox_Privileges->Text.Pos("OWNER") > 0)
        hbTapiLine1->Privileges->Owner = true;
      if (ComboBox_Privileges->Text.Pos("MONITOR") > 0)
        hbTapiLine1->Privileges->Monitor = true;
    };
    hbTapiLine1->Active = true;
    Label_Privileges->Caption = "Privileges: " + hbTapiLine1->Privileges->Text;
  };
  UpdateCalls();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormCreate(TObject *Sender)
{
  ComboBoxDevice->Items = hbTapiLine1->DeviceList;
  ComboBoxDevice->Items->Insert(0, "- none -");
  ComboBox_Privileges->ItemIndex = 0;
  ComboBoxDevice->ItemIndex = 0;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::UpdateCalls()
{
  int i; TListItem* li; ThbTapiCall* c;
  c = NULL;

  for (i=0;i<ListView1->Items->Count;i++)
  {
    if ((ListView1->Items->Item[i]->Selected || ListView1->Items->Item[i]->Focused)
         && (hbTapiLine1->Calls->IndexOf(ListView1->Items->Item[i]->Data) >= 0))
    {
      c = (ThbTapiCall*)ListView1->Items->Item[i]->Data;
      break;
    };
  };

  for (i=0;i<hbTapiLine1->Calls->Count;i++)
  {
    if (i < ListView1->Items->Count)
      li = ListView1->Items->Item[i];
    else
      li = ListView1->Items->Add();

    while (li->SubItems->Count < ListView1->Columns->Count)
      li->SubItems->Add("");

    li->Caption = "0x" + IntToHex((int)hbTapiLine1->Calls->Items[i],8);
    li->Data = hbTapiLine1->Calls->Items[i];
    li->SubItems->Strings[0] = IntToStr((int)hbTapiLine1->Calls->Items[i]->CallID);

    if (hbTapiLine1->Calls->Items[i]->IsOutgoing())
      li->SubItems->Strings[1] = hbTapiLine1->Calls->Items[i]->CalledID->Address;
    else
      li->SubItems->Strings[1] = hbTapiLine1->Calls->Items[i]->CallerID->Address;
    li->SubItems->Strings[2]   = hbTapiLine1->Calls->Items[i]->StateText();
    li->SubItems->Strings[3]   = hbTapiLine1->Calls->Items[i]->PrivilegeText();

    li->ImageIndex    = 0;
    switch (hbTapiLine1->Calls->Items[i]->State)
    {
      case LINECALLSTATE_IDLE               : li->ImageIndex = 1; break;
      case LINECALLSTATE_OFFERING           : li->ImageIndex = 4; break;
      case LINECALLSTATE_ACCEPTED           :; break;
      case LINECALLSTATE_DIALTONE           :
      case LINECALLSTATE_DIALING            :
      case LINECALLSTATE_RINGBACK           :
      case LINECALLSTATE_PROCEEDING         : li->ImageIndex = 2; break;
      case LINECALLSTATE_BUSY               : li->ImageIndex = 8; break;
      case LINECALLSTATE_SPECIALINFO        :; break;
      case LINECALLSTATE_CONNECTED          : li->ImageIndex = 3; break;
      case LINECALLSTATE_CONFERENCED        : li->ImageIndex = 7; break;
      case LINECALLSTATE_ONHOLD             :
      case LINECALLSTATE_ONHOLDPENDCONF     :
      case LINECALLSTATE_ONHOLDPENDTRANSFER : li->ImageIndex = 5; break;
      case LINECALLSTATE_DISCONNECTED       : li->ImageIndex = 6; break;
    }
  }

  while (ListView1->Items->Count > hbTapiLine1->Calls->Count)
    ListView1->Items->Delete(ListView1->Items->Count-1);

  for (i=0;i<ListView1->Items->Count;i++)
  {
    if (ListView1->Items->Item[i]->Data == c)
      ListView1->Selected = ListView1->Items->Item[i];
  }
  if (!ListView1->Selected && (ListView1->Items->Count > 0))
    ListView1->Selected = ListView1->Items->Item[0];

  if (!c && (ListView1->Items->Count > 0))
    c = (ThbTapiCall*)ListView1->Items->Item[0]->Data;

  if (c)
    Label_CallFeature->Caption = "CallFeatures: " + c->FeaturesText();
  else
    Label_CallFeature->Caption = "";

  ListView1->Update();
}
void __fastcall TForm1::hbTapiLine1CallState(ThbTapiLine *Sender,
      ThbTapiCall *Call, DWORD CallState)
{
  ListBox1->Items->Add("::OnCallState " + Call->StateText() + " (Call = 0x" + IntToHex((int)Call, 8) + ")");
  UpdateCalls();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::hbTapiLine1CallDeallocate(ThbTapiLine *Sender,
      ThbTapiCall *Call)
{
  UpdateCalls();
}
//---------------------------------------------------------------------------
//---------------------------------------------------------------------------
ThbTapiCall* __fastcall TForm1::GetCall()
{
  if (hbTapiLine1->Calls->Count > 0)
  {
    if (ListView1->Selected)
    {
      if (hbTapiLine1->Calls->IndexOf((ThbTapiCall*)ListView1->Selected->Data) == -1)
        return NULL;
      else
        return (ThbTapiCall*)ListView1->Selected->Data;
    }
  }
  return NULL;
}
void __fastcall TForm1::Button2Click(TObject *Sender)
{
  ThbTapiCallParams* cp;

  if ((InputQuery("MakeCall", "Phone Number", FPhoneNumber) == false) || (FPhoneNumber == ""))
     exit;

  // here we use the ThbTapiCallParams object instead of ThbTapiLine->CallParams
  cp = new ThbTapiCallParams();
  try
  {
    cp->AddressMode   = LINEADDRESSMODE_DIALABLEADDR;
    cp->AddressID     = ComboBoxAddress->ItemIndex;
    cp->MediaMode     = LINEMEDIAMODE_INTERACTIVEVOICE;
    cp->Flags         = 0;
    cp->TargetAddress = FPhoneNumber;
    hbTapiLine1->MakeCall(FPhoneNumber, 0, cp->TapiData);
  }
  __finally
  {
    delete cp;
  }

}
//---------------------------------------------------------------------------
void __fastcall TForm1::Button4Click(TObject *Sender)
{
  ThbTapiCall* c;

  c = GetCall();
  if (!c)
    return;

  if ((InputQuery("Dial", "Phone Number", FPhoneNumber) == false) || (FPhoneNumber == ""))
    return;

  hbTapiLine1->CallParams->AddressID = ComboBoxAddress->ItemIndex;
  c->Dial(FPhoneNumber);
        
}
//---------------------------------------------------------------------------
void __fastcall TForm1::Button5Click(TObject *Sender)
{
  ThbTapiCall* c;
  c = GetCall();
  if (!c)
    return;
  c->Answer("");
        
}
//---------------------------------------------------------------------------
void __fastcall TForm1::Button6Click(TObject *Sender)
{
  ThbTapiCall* c;
  c = GetCall();
  if (!c)
    return;
  c->Drop("");


}
//---------------------------------------------------------------------------
void __fastcall TForm1::ButtonSetCallDataClick(TObject *Sender)
{
  ThbTapiCall* c;
  c = GetCall();
  if (!c)
    return;
  c->SetCallData((void*)Edit1->Text.c_str(), Edit1->Text.Length()+1);

}
//---------------------------------------------------------------------------
void __fastcall TForm1::ButtonGetCallDataClick(TObject *Sender)
{
  ThbTapiCall* c; void* p;
  c = GetCall();
  if (!c)
    return;

  if (c->GetCallData(p) > 0 )
  try
  {
    Edit2->Text = PChar(p);
  }
  __finally
  {
    free(p);
  };

}
//---------------------------------------------------------------------------

File: hbTapi Components/Delphi Samples/Enterprise Edition/CallData/uMain.pas
{-------------------------------------------------------------------------
   Copyright by Haeger + Busch, Germany                / >>>>>>>>> /-----
   Ingenieurbuero fuer Kommunikationslösungen         / <<<<<<<<< /
 ----------------------------------------------------/ >>>>>>>>> /
   Homepage  : http://www.hbTapi.com
   EMail     : info@hbTapi.com
   Package   : hbTapi Components
 -------------------------------------------------------------------------}
unit uMain;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, ExtCtrls, hbTAPI, hbTapiUtils, hbTapiApi, ComCtrls, ImgList, Menus,
  hbComm;

type
  TForm1 = class(TForm)
    hbTapiLine1: ThbTapiLine;
    GroupBox1: TGroupBox;
    ComboBoxDevice: TComboBox;
    CheckBoxActive: TCheckBox;
    Label1: TLabel;
    GroupBox2: TGroupBox;
    ListView1: TListView;
    PopupMenu1: TPopupMenu;
    Answer1: TMenuItem;
    Drop1: TMenuItem;
    MakeCall1: TMenuItem;
    Dial1: TMenuItem;
    Button2: TButton;
    GroupBox3: TGroupBox;
    ButtonSetCallData: TButton;
    GroupBox4: TGroupBox;
    ButtonGetCallData: TButton;
    Button4: TButton;
    Button5: TButton;
    Button6: TButton;
    UpdateCalls1: TMenuItem;
    ImageList1: TImageList;
    Label3: TLabel;
    Edit1: TEdit;
    Label2: TLabel;
    Edit2: TEdit;
    Label4: TLabel;
    N1: TMenuItem;
    Label_Privileges: TLabel;
    ComboBox_Privileges: TComboBox;
    Label5: TLabel;
    procedure CheckBoxActiveClick(Sender: TObject);
    procedure ComboBoxDeviceChange(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure DoMakeCall(Sender: TObject);
    procedure DoSwapHold(Sender: TObject);
    procedure hbTapiLine1CallState(Sender: ThbTapiLine; Call: ThbTapiCall;
      CallState: Cardinal);
    procedure hbTapiLine1CallDeallocate(Sender: ThbTapiLine; Call: ThbTapiCall);
    procedure hbTapiLine1NewCall(Sender: ThbTapiLine; Call: ThbTapiCall);
    procedure DoDropCall(Sender: TObject);
    procedure DoDial(Sender: TObject);
    procedure DoAnswerCall(Sender: TObject);
    procedure DoUpdateCalls(Sender: TObject);
    procedure ListView1Click(Sender: TObject);
    procedure ButtonSetCallDataClick(Sender: TObject);
    procedure ButtonGetCallDataClick(Sender: TObject);
  private
    FPhonenumber : String;
    function GetCall: ThbTapiCall;
    procedure UpdateCalls;
  public
    { Public declarations }
  end;

var Form1:        TForm1;

implementation

{$R *.DFM}

procedure TForm1.CheckBoxActiveClick(Sender: TObject);
begin
  if not CheckBoxActive.Checked then
    begin
      if hbTapiLine1.Active then
      begin
        hbTapiLine1.Active := False;
        Label_Privileges.Caption := '';
      end;
    end
  else
    begin
      if not hbTapiLine1.Active then
      begin
        if Pos('AutoSelect', ComboBox_Privileges.Text) > 0 then
          hbTapiLine1.Privileges.AutoSelect := True
        else  
        begin
          hbTapiLine1.Privileges.None := True;
          hbTapiLine1.Privileges.AutoSelect := False;
          if Pos('OWNER', ComboBox_Privileges.Text) > 0 then
            hbTapiLine1.Privileges.Owner := True;
          if Pos('MONITOR', ComboBox_Privileges.Text) > 0 then
            hbTapiLine1.Privileges.Monitor := True;
        end;
        hbTapiLine1.Active := True;
        Label_Privileges.Caption := Format('Privileges: %s', [hbTapiLine1.Privileges.Text]);
      end;
    end;
  UpdateCalls;
end;

procedure TForm1.ComboBoxDeviceChange(Sender: TObject);
begin
  if hbTapiLine1.Active then
  begin
    CheckBoxActive.Checked := False;
    hbTapiLine1.Active := FALSE;
  end;
  hbTapiLine1.DeviceID := ComboBoxDevice.ItemIndex;
end;

procedure TForm1.FormCreate(Sender: TObject);
begin
  ComboBoxDevice.Items := hbTapiLine1.DeviceList;
  ComboBox_Privileges.ItemIndex := 0;
  if ComboBoxDevice.Items.Count > 0 then
    ComboBoxDevice.Text := ComboBoxDevice.Items[0];
  ComboBoxDevice.ItemIndex := 0;
  ComboBoxDeviceChange(ComboBoxDevice);
end;

procedure TForm1.UpdateCalls;
var i: integer; li: TListItem; c: ThbTapiCall;
begin
  c := nil;
  for i := 0 to ListView1.Items.Count -1 do
  begin
    if (ListView1.Items[i].Selected or ListView1.Items[i].Focused) and (hbTapiLine1.Calls.IndexOf(ListView1.Items[i].Data) >= 0) then
    begin
      c := ThbTapiCall(ListView1.Items[i].Data);
      break;
    end;
  end;

  if c <> nil then
    Label3.Caption := 'CallFeatures: ' + c.FeaturesText
  else
    Label3.Caption := '';

  with hbTapiLine1 do
  begin
    for i:= 0 to Calls.Count-1 do
    begin
      if i < ListView1.Items.Count then
        li := ListView1.Items[i]
      else
        li := ListView1.Items.Add;

      while li.SubItems.Count < ListView1.Columns.Count do
        li.SubItems.Add('');

      li.Caption       := Format('$%x',[integer(Calls[i])]);
      li.Data          := Calls[i];
      li.SubItems[0]   := inttostr(Calls[i].CallID);

      if Calls[i].IsOutgoing then
        li.SubItems[1] := Calls[i].CalledID.Address
      else
        li.SubItems[1] := Calls[i].CallerID.Address;
      li.SubItems[2]   := Calls[i].StateText;
      li.SubItems[3]   := Calls[i].PrivilegeText;

      li.ImageIndex    := 0;
      case Calls[i].State of
        LINECALLSTATE_IDLE               : li.ImageIndex := 1;
        LINECALLSTATE_OFFERING           : li.ImageIndex := 4;
        LINECALLSTATE_ACCEPTED           :;
        LINECALLSTATE_DIALTONE,
        LINECALLSTATE_DIALING,
        LINECALLSTATE_RINGBACK,
        LINECALLSTATE_PROCEEDING         : li.ImageIndex := 2;
        LINECALLSTATE_BUSY               : li.ImageIndex := 8;
        LINECALLSTATE_SPECIALINFO:;
        LINECALLSTATE_CONNECTED          : li.ImageIndex := 3;
        LINECALLSTATE_CONFERENCED        : li.ImageIndex := 7;
        LINECALLSTATE_ONHOLD,
        LINECALLSTATE_ONHOLDPENDCONF,
        LINECALLSTATE_ONHOLDPENDTRANSFER : li.ImageIndex := 5;
        LINECALLSTATE_DISCONNECTED       : li.ImageIndex := 6;
      end;
    end;
    while ListView1.Items.Count > Calls.Count do
      ListView1.Items[ListView1.Items.Count-1].Delete;
    for i := 0 to ListView1.Items.Count -1 do
    begin
      if ListView1.Items[i].Data = c then
      begin
        ListView1.Selected := ListView1.Items[i];
      end;
    end;
    if (ListView1.Selected = nil) and (ListView1.Items.Count > 0) then
      ListView1.Selected := ListView1.Items[0];
  end;
  ListView1.Update;
end;

function TForm1.GetCall: ThbTapiCall;
begin
  result := nil;

  if (ListView1.Selected = nil) and (ListView1.Items.Count >0) then
    ListView1.Selected := ListView1.Items[0];

  if ListView1.Selected = nil then
  begin
    MessageDlg('No call available.', mtInformation, [mbOk], 0);
    exit;
  end;

  result := ThbTapiCall(ListView1.Selected.Data);

  if hbTapiLine1.Calls.IndexOf(result) = -1 then
      result := nil;
end;

procedure TForm1.DoMakeCall(Sender: TObject);
begin
  if InputQuery('Dial to', 'Phonenumber', FPhonenumber) then
    hbTapiLine1.MakeCall(FPhonenumber);
end;

procedure TForm1.DoDial(Sender: TObject);
var c: ThbTapiCall;
begin
  c := GetCall;
  if c = nil then
    exit;
  if InputQuery('Dial to', 'Phonenumber', FPhonenumber) then
    c.Dial(FPhonenumber);
end;

procedure TForm1.DoSwapHold(Sender: TObject);
var i: integer; c1, c2: ThbTapiCall;
begin
  c1:= GetCall;
  if c1 = nil then exit;
  c2:= nil;
  for i := 0 to hbTapiLine1.Calls.Count -1 do
  begin
    if (hbTapiLine1.Calls[i] <> c1) and
       ((c1.IsConnected and hbTapiLine1.Calls[i].IsHeld) or (c1.IsHeld and hbTapiLine1.Calls[i].IsConnected)) then
    begin
      c2 := hbTapiLine1.Calls[i];
    end;
  end;
  if (c2 <> nil) then
  begin
    if c1.IsConnected then
      c1.SwapHold(c2)
    else
      c2.SwapHold(c1);
  end;
end;

procedure TForm1.hbTapiLine1CallState(Sender: ThbTapiLine; Call: ThbTapiCall;
  CallState: Cardinal);
begin
  UpdateCalls;
end;

procedure TForm1.hbTapiLine1CallDeallocate(Sender: ThbTapiLine;
  Call: ThbTapiCall);
begin
  UpdateCalls;
end;

procedure TForm1.hbTapiLine1NewCall(Sender: ThbTapiLine; Call: ThbTapiCall);
begin
  UpdateCalls;
end;

procedure TForm1.DoDropCall(Sender: TObject);
var c: ThbTapiCall;
begin
  c := GetCall;
  if c = nil then exit;
  c.Drop;
end;

procedure TForm1.DoAnswerCall(Sender: TObject);
var c: ThbTapiCall;
begin
  c := GetCall;
  if c = nil then exit;
  c.Answer;
end;

procedure TForm1.DoUpdateCalls(Sender: TObject);
begin
  UpdateCalls;
end;

procedure TForm1.ListView1Click(Sender: TObject);
begin
  UpdateCalls;
end;

procedure TForm1.ButtonSetCallDataClick(Sender: TObject);
var c: ThbTapiCall;
begin
  c := GetCall;
  if c = nil then exit;
  c.SetCallData(PAnsiChar(AnsiString(Edit1.Text)), length(Edit1.Text)+1);
end;

procedure TForm1.ButtonGetCallDataClick(Sender: TObject);
var c: ThbTapiCall; p: Pointer;
begin
  c := GetCall;
  if c = nil then exit;
  if c.GetCallData(p) > 0 then
  try
    Edit2.Text := PAnsiChar(p);
  finally
    FreeMem(p);
  end;
end;

end.


